baukit/nethook.py

```diff
@@ -58,42 +62,84 @@ def __init__(
+                            key: recursive_copy(
+                                value, clone=clone, detach=detach, retain_grad=False
+                            )
+                            for key, value in kwargs.items()
+                        }
+                        retainer.input = inputs
+                    else:
+                        if len(args) != 0:
+                            # the main input is often the first argument
+                            # TODO: is this **always** true?
+                            inputs = args[0] if len(args) == 1 else args
+                        else:
+                            assert (
+                                input_key is not None
+                            ), "all arguments are passed as kwargs. Please specify `input_key`"
+                            assert (
+                                input_key in kwargs
+                            ), f"input_key {input_key} not found in kwargs"
+                            inputs = kwargs[input_key]
+
+                        retainer.input = recursive_copy(
+                            inputs,
+                            clone=clone,
+                            detach=detach,
+                            retain_grad=False,  # retain_grad applies to output only.
+                        )
+
+                # ----------------------------------
+                # call the actual forward function
+                output = forward_fn(*args, **kwargs)
+                # ----------------------------------
+
+                if edit_output:
+                    output = invoke_with_optional_args(
+                        edit_output, output=output, layer=self.layer, inputs=inputs
+                    )
+
+                if retain_output:
+                    retainer.output = recursive_copy(
+                        output, clone=clone, detach=detach, retain_grad=retain_grad
+                    )
+                    # When retain_grad is set, also insert a trivial
+                    # copy operation.  That allows in-place operations
+                    # to follow without error.
+                    if retain_grad:
+                        output = recursive_copy(
+                            retainer.output, clone=True, detach=False
+                        )
+                return output
+
+            return wrap_forward
+
         self.layer = layer
         if layer is not None:
-            module = get_module(module, layer)
+            self.module = get_module(module, layer)
+            self.hook = self.module.forward
+            self.module.forward = retain_hook(self.module.forward)
 
-        def retain_hook(m, inputs, output):
-            if edit_output:
-                output = invoke_with_optional_args(
-                    edit_output, output=output, layer=self.layer, inputs=inputs
-                )
-            if retain_input:
-                retainer.input = recursive_copy(
```

Comment:
how bad would it be to simply add something like
```
retainer.input_kw = recursive_copy(kwargs)
```

and then skip all the extra arguments on Trace?  (i.e., no `input_key` or `input_all_keys` arguments.)